src/circuit_builder.rs

```diff
@@ -0,0 +1,304 @@
+//! Tools & traits for PLONK circuits
+
+use crate::commitment_scheme::kzg10::PublicParameters;
+use crate::constraint_system::StandardComposer;
+use crate::proof_system::{Proof, ProverKey, VerifierKey};
+use anyhow::{Error, Result};
+use dusk_bls12_381::Scalar as BlsScalar;
+use dusk_jubjub::{AffinePoint as JubJubAffine, Scalar as JubJubScalar};
+
+/// Circuit inputs
+#[derive(Debug, Clone, Copy)]
+pub struct CircuitInputs<'a> {
+    bls_scalars: &'a [BlsScalar],
+    jubjub_scalars: &'a [JubJubScalar],
+    jubjub_affines: &'a [JubJubAffine],
+}
+
+/// Public Input
+#[derive(Debug, Copy, Clone)]
+pub enum PublicInput {
```

Comment:
We should have mechanisms to serialize/deserialize a combination of `PublicInput` so these can be trivially broadcasted

Reply:
If they are public there should be no need to broadcast them no?

Reply:
They must always be stored on the chain or sent via the network with the proof.

Reply:
Ye but they aren't sent like that. You should just send the values. not the positions that they occupy inside the PI vector.